### ISSUE TITLE
minification issues

### ISSUE THREAD

**Issue by markuz-brasil** [angular/di.js#95]
First of all, I'm using 6to5 instead of traceur and mocha/chai/sinon instead of karma/jasmine. (and yeah, I ported the tests)

Flame war apart, the tests shouldn't fail. Which doesn't, unless it is minified. (And I didn't bother trying with minified traceur output...)

I pinpointed the issue to the class' context being somehow set to `undefined` in the minified form. So things like `this.list` fails because `this` is `undefined`.

Here is my project's [git](https://github.com/markuz-brasil/runtime#1e67a55cffec352cf9ef48fc26344ec5f61f68abbrasil.github.io/runtime/build/) (minified) and all [**passing tests**](http://markuz-brasil.github.io/runtime/dev/)

I'm not clear where the issue actually is. I'm 99% certain that it isn't the testing framework. #93 doesn't fix it either. I'm left with this 6to5/6to5#343, a bug on uglify or on the di framework itself. (or some combination of them)

My work around is something like this:
instead of writing this:

``` javascript
class UserList {}

class UserController {
  constructor(list) {
    this.list = list
  }
}
annotate(UserController, new Inject(UserList))
```

write this:

``` javascript
class UserList {}

class UserController {
  constructor(list) {
    this.list = list
  }
}

// add a factory around a class :/
annotate(UserFactory, new Inject(UserList))
function UserFactory (list) {
  return new UserController(list)
}
```

not ideal, but at least it gets the job done

**Comment by tusharmath** [comment]
@markuz-brasil How do you know this is not getting fixed with #93

**Comment by markuz-brasil** [comment]
I took a look at the code, copied and pasted.
But I'm 100% convinced that the `isClass` and uglify are at fault.

I found out that uglifyy doesn't support Function.name [link](
http://lisperator.net/uglifyjs/mangle)

I think that there is no way to distinguish a class from a function in ES5
(maybe ES6 too). So `class Foo {}` and `function Foo(){}` are exactly the
same.

I feel like the solution is with annotations. Maybe a `@Factory` and
`isClass`check for that with `hasAnnotatiom(FnOrClass, Factory)`

**Comment by caitp** [comment]
> I think that there is no way to distinguish a class from a function in ES5
(maybe ES6 too).

No, in ES6 they are deserialized as a class declaration or class expression

**Comment by markuz-brasil** [comment]
Cool. Good to know. Thanks

**Comment by tusharmath** [comment]
@markuz-brasil Actually #93 is a workaround so that the class doesn't throw up errors.
In most real world scenarios every class would have some method. So if you update your code like this - 


```js

class UserController {
  constructor(list) {
    this.list = list
  }

  //Add a method to the prototype chain
  methodA(){

  }
}
```

It should work.

**Comment by iammerrick** [comment]
Indeed, here is an example project: https://github.com/iammerrick/webpack-jsx-uglify-bug

**Comment by vojtajina** [comment]
There is no way to distinguish a class from a function, di.js uses the function/class name - if it starts with an uppercase letter, it's treated as a class. Sounds like the minifier changes class names into lowercase letters.

We can't treat everything as a class (call it with `new` operator), because if a non-class function returns a primitive value, this value wouldn't be used (the new object is used instead) and that's incorrect behavior.

We can add additional annotation such as `@ClassProvider` or `@FactoryProvider`.

Any other ideas?

**Comment by iammerrick** [comment]
I think the additional annotation is our best option.

**Comment by markuz-brasil** [comment]
I also agree.

Maybe something like @Constructor as a hint for using ClassProvider and
the FactoryProvider as default way.

**Comment by tusharmath** [comment]
Its just getting too verbose. Why can't we follow a convention as opposed to configuration?

Like any function with a name can be considered as a class otherwise its a factory.

**Comment by iammerrick** [comment]
This is also a problem with @InjectLazy.

**Comment by tusharmath** [comment]
Is there a drawback of using the annotation approach?

**Comment by iammerrick** [comment]
Addressing this with annotations here: https://github.com/angular/di.js/pull/96/files
